webhook/strict.go

```diff
@@ -0,0 +1,313 @@
+		}
+	}
+
+	oldField := oldValue.FieldByName(name)
+	if !oldField.IsValid() || oldField.Kind() != reflect.Struct {
+		return &apis.FieldError{
+			Message: "failed to inspect old object",
+			Paths:   []string{apis.CurrentField},
+		}
+	}
+
+	if oldField.Type() != resField.Type() {
+		return &apis.FieldError{
+			Message: "failed to match filed type",
+			Paths:   []string{apis.CurrentField},
+		}
+	}
+
+	return strictValidateUpdateReflected(oldField, resField)
+}
+
+func strictValidateUpdateReflected(oldField, resField reflect.Value) *apis.FieldError {
+	if !resField.IsValid() || resField.Kind() != reflect.Struct {
+		return &apis.FieldError{
+			Message: "failed to inspect new object",
+			Paths:   []string{apis.CurrentField},
+		}
+	}
+
+	if !oldField.IsValid() || oldField.Kind() != reflect.Struct {
+		return &apis.FieldError{
+			Message: "failed to inspect old object",
+			Paths:   []string{apis.CurrentField},
+		}
+	}
+
+	if oldField.Type() != resField.Type() {
+		return &apis.FieldError{
+			Message: "failed to match filed type",
+			Paths:   []string{apis.CurrentField},
+		}
+	}
+
+	var errs *apis.FieldError
+
+	// for each field,
+	for i := 0; i < resField.NumField(); i++ {
+		f := resField.Type().Field(i)
+		if v := resField.Field(i); v.IsValid() {
+			if o := oldField.Field(i); o.IsValid() {
+				if strings.HasPrefix(f.Name, "Deprecated") {
+					if differ(v, o) {
+						// TODO: add a field error.
+						errs = errs.Also(&apis.FieldError{
+							Message: "deprecated field updated",
+							Paths:   []string{f.Name},
+						})
+						continue
+					}
+
+					// Can be deleted.
+					if nonZero(v) {
+						errs = errs.Also(&apis.FieldError{
+							Message: "deprecated field set",
+							Paths:   []string{f.Name},
+						})
+
+					}
+				}
+
+				switch v.Kind() {
+				case reflect.Ptr:
+					if v.IsNil() {
+						continue
+					}
+					if o.IsNil() {
+						errs = errs.Also(
+							strictValidateCreateReflected(v.Elem()).ViaField(f.Name),
+						)
+					} else {
+						errs = errs.Also(
+							strictValidateUpdateReflected(o.Elem(), v.Elem()).ViaField(f.Name),
+						)
+					}
+
+				case reflect.Struct:
+					errs = errs.Also(
+						strictValidateUpdateReflected(o, v).ViaField(f.Name),
+					)
+
+				// TODO: because slices have no order, I am not sure how to test for this.
+				//case reflect.Slice, reflect.Array:
+
+				case reflect.Map:
+					it := v.MapRange()
+					for it.Next() {
+						key := it.Key()
+						res := it.Value()
+						old := o.MapIndex(key)
+						errs = errs.Also(
+							strictValidateUpdateReflected(old, res).ViaFieldKey(f.Name, key.String()),
+						)
+					}
+				}
+
+			}
+		}
+	}
+	return errs
+}
+
+// nonZero returns true if a != 0
+func nonZero(a reflect.Value) bool {
+	switch a.Kind() {
+	case reflect.Ptr:
+		if a.IsNil() {
+			return false
+		}
+		return nonZero(a.Elem())
+
+	case reflect.Map, reflect.Slice, reflect.Array:
+		if a.IsNil() {
+			return false
+		}
```

Comment:
empty as well?

Reply:
I am counting empty list as invalid. Not allowed to set the deprecated field.